nomad/state/state_store.go

```diff
@@ -926,6 +934,95 @@ func appendNodeEvents(index uint64, node *structs.Node, events []*structs.NodeEv
 	}
 }
 
+// upsertNodeCSIPlugins indexes csi plugins for volume retrieval, with health. It's called
+// on upsertNodeEvents, so that event driven health changes are updated
+func upsertNodeCSIPlugins(txn *memdb.Txn, node *structs.Node, index uint64) error {
+	if len(node.CSIControllerPlugins) == 0 && len(node.CSINodePlugins) == 0 {
+		return nil
+	}
+
+	loop := func(info *structs.CSIInfo) error {
+		raw, err := txn.First("csi_plugins", "id", info.PluginID)
+		if err != nil {
+			return fmt.Errorf("csi_plugin lookup error: %s %v", info.PluginID, err)
+		}
+
+		var plug *structs.CSIPlugin
+		if raw != nil {
+			plug = raw.(*structs.CSIPlugin).Copy(index)
+		} else {
+			plug = structs.NewCSIPlugin(info.PluginID, index)
+		}
+
+		plug.AddPlugin(node.ID, info, index)
+
+		err = txn.Insert("csi_plugins", plug)
+		if err != nil {
+			return fmt.Errorf("csi_plugins insert error: %v", err)
+		}
+
+		return nil
+	}
+
+	for _, info := range node.CSIControllerPlugins {
+		err := loop(info)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, info := range node.CSINodePlugins {
+		err := loop(info)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deleteNodeCSIPlugins cleans up CSIInfo node health status, called in DeleteNode
+func deleteNodeCSIPlugins(txn *memdb.Txn, node *structs.Node, index uint64) error {
+	if len(node.CSIControllerPlugins) == 0 && len(node.CSINodePlugins) == 0 {
+		return nil
+	}
+
+	names := map[string]struct{}{}
+	for _, info := range node.CSIControllerPlugins {
+		names[info.PluginID] = struct{}{}
+	}
+	for _, info := range node.CSINodePlugins {
+		names[info.PluginID] = struct{}{}
+	}
+
+	for id := range names {
+		raw, err := txn.First("csi_plugins", "id", id)
+		if err != nil {
+			return fmt.Errorf("csi_plugins lookup error %s: %v", id, err)
+		}
+		if raw == nil {
+			continue // FIXME error?
```

Comment:
It's not clear to me what case this is (does `txn.First` return an error if the key is missing?), but if we're not sure I'd rather return the error now and have it to deal with, rather than forget about it later.

Reply:
`txn.First` does not return an error for a missing record, just a nil entry. I like that suggestion, though